src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java

```diff
@@ -57,50 +60,92 @@ protected StructuredRecord computeNext() {
-        String fieldName = field.getName();
-        Object value = cur._2().get(fieldName);
-        if (blacklist.contains(fieldName)) {
-          continue;
-        }
-
-        // check if there is difference between next record and cur record
-        Object nextVal = next._2().get(fieldName);
-        if ((nextVal == null) != (value == null) || (value != null && !value.equals(nextVal))) {
-          isDiff = true;
-          break;
-        }
-      }
-
-      if (!isDiff) {
-        prev = cur;
-        cur = next != null ? next : records.next();
-        next = records.hasNext() ? records.next() : null;
-      }
+    if (prev == null) {
+      cur = deduplicateNext();
+    } else {
+      cur = next;
     }
+    next = records.hasNext() ? deduplicateNext() : null;
 
     // if key changes, clean up the table to free memory
     if (prev != null && !prev._1().equals(cur._1())) {
-      valTable.row(prev._1().getKey()).clear();
+      valTable.row(prev._1()).clear();
     }
 
-    return computeRecord(cur._1().getKey(),
+    return computeRecord(cur._1(),
                          prev != null && prev._1().equals(cur._1()) ? prev._2() : null,
                          cur._2(),
                          next != null && next._1().equals(cur._1()) ? next._2() : null);
   }
 
-  private StructuredRecord computeRecord(Object key, @Nullable StructuredRecord prev, StructuredRecord cur,
+  /**
+   * Find the next record that will not get deduplicated and get rid of the other duplicated records.
+   *
+   * For example, consider the following example:
+   * 1. id: 1, name: john, isTarget: false
+   * 2. id: 1, name: john, isTarget: true
+   * 3. id: 1, name: john, isTarget: false
+   * 4. id: 1, name: sam, isTarget: false
+   * 5. id: 1, name: sam, isTarget: false
+   *
+   * After this method, the records are processed like this:
+   * 1. id: 1, name: john, isTarget: false (get ignored since target is false and it is same as #2)
+   * 2. id: 1, name: john, isTarget: true (will be returned since target is true)
+   * 3. id: 1, name: john, isTarget: false (get ignored since target is false and it is same as #2)
+   * 4. id: 1, name: sam, isTarget: false (iterator.peek() now points to it since it is a candiate for the
+   *                                       next non-dedup record)
+   * 5. id: 1, name: sam, isTarget: false (will not get processed in the method)
+   */
+  private Tuple2<SCD2Key, StructuredRecord> deduplicateNext() {
+    Tuple2<SCD2Key, StructuredRecord> currentElement = records.next();
+    Tuple2<SCD2Key, StructuredRecord> result = currentElement;
+    boolean findResult = false;
+
+    while (conf.deduplicate() && records.hasNext()) {
+      Tuple2<SCD2Key, StructuredRecord> nextElement = records.peek();
+      if (isTarget(currentElement._2())) {
+        findResult = true;
+      }
+
+      // if the current is different from next or if the next is from target table, we can stop deduplicating
```

Comment:
next is not expected to be a target right? This is just in case there happen to be dupes in the target?

Reply:
Yes, this is just to guard against in case the target contains some duplicates. A unit test was also added for this.